src/k2/persistence/logStream/LogStream.h

```diff
@@ -0,0 +1,163 @@
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#pragma once
+
+#include <k2/transport/PayloadSerialization.h>
+#include <seastar/core/sharded.hh>
+#include <k2/transport/Payload.h>
+#include <k2/transport/Status.h>
+#include <k2/common/Common.h>
+#include <k2/config/Config.h>
+#include <k2/cpo/client/CPOClient.h>
+#include <k2/transport/BaseTypes.h>
+#include <k2/transport/TXEndpoint.h>
+#include <k2/persistence/plog_client/PlogClient.h>
+
+
+namespace k2 {
+
+namespace log {
+inline thread_local k2::logging::Logger lgbase("k2::logstream_base");
+}
+
+namespace log {
+inline thread_local k2::logging::Logger mdmgr("k2::metadata_manager");
+}
+
+namespace log {
+inline thread_local k2::logging::Logger logstream("k2::logstream");
+}
+
+class LogStreamBase;
+class LogStream;
+class MetadataMgr;
+
+// a base class that provide operations to handle the plogs. it will be used by both log stream and metadata manager
+class LogStreamBase{
+private:
+    // to store the information of each plog id used by the log stream
+    struct PlogInfo {
+        uint32_t currentOffset;
+        bool sealed;
+        uint32_t index;
+    };
+
+    struct MetadataElement{
+        String plogId;
+        uint32_t start_offset;
+        uint32_t size;
+    };
+public:
+    LogStreamBase();
+    ~LogStreamBase();
+    
+    // create a log stream base
+    seastar::future<> create();
+
+    // write data to the the current plog, return the latest Plog ID and latest offset
+    seastar::future<std::pair<String, uint32_t> > append(Payload payload);
+
+    // write data to the the current plog with the start offset, return the latest Plog ID and latest offset
+    seastar::future<std::pair<String, uint32_t> > append(Payload payload, String plogId, uint32_t offset);
+
+    // read all the data from this log stream
+    seastar::future<std::vector<Payload> > read(String start_plogId, uint32_t start_offset, uint32_t size);
+
+    // reload the _usedPlogIdVector and _usedPlogInfo for replay purpose
+    seastar::future<Status> reload(std::vector<dto::MetadataRecord> records);
+
+    // init the plog client client
+    seastar::future<> init_plog_client(String cpo_url, String persistenceClusrerName);
+
+    // obtain the target plog status
+    seastar::future<std::tuple<Status, std::tuple<uint32_t, bool>>> get_plog_status(String plogId);
+private:
+    PlogClient _client;
+    // the maximum size of each plog
+    constexpr static uint32_t PLOG_MAX_SIZE = 16 * 1024 * 1024;
+    // how many redundant plogs it will create in advance 
+    constexpr static uint32_t PLOG_POOL_SIZE = 1;
+
+    // whether this log stream base has been created 
+    bool _create = false; 
+
+    // The vector to store the redundant plog Id
+    std::vector<String> _redundantPlogPool;
+    // The vector to store the used plog Id
```

Comment:
One map should be enough for both purpose.

Reply:
fixed